### DISCUSSION TITLE
out-of-tree serial driver get No SOURCES warning

### DISCUSSION THREAD

**Discussion by foxriver1025** [zephyrproject-rtos/zephyr#71734] [General] [answered]
https://github.com/zephyrproject-rtos/zephyr/blob/1c3edca8d89c7f66d2c0c0f5bec631517a6fe468/drivers/sensor/CMakeLists.txt#L69C25-L69C36

out-of-tree driver, add a new serial driver, get warning as this:
  No SOURCES given to Zephyr library: drivers__serial

  Excluding target from build.
serial driver does not add "zephyr_library_property(ALLOW_EMPTY TRUE)"

**Comment by gramsay0** [comment, marked as answer]
Try using `zephyr_library_amend()` rather than `zephyr_library()` in out-of-tree cmake.
This amends the Zephyr `drivers__serial` library with your out-of-tree additions, rather than creating a new library.
Otherwise one of the two libraries will likely be empty and raise that warning.
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/extensions.cmake#L461

**Comment by foxriver1025** [comment]
@gramsay0 Thanks for your reply, I try to use zephyr_library_amend() rather than zephyr_library() in out-of-tree cmake, but a compilation error occurred:

```
CMake Error at /home/zhoubin/code/m88star5_work/zephyr/cmake/modules/extensions.cmake:482 (target_sources):
  Cannot specify sources for target "drivers__serial__atcuart100" which is
  not built by this project.
Call Stack (most recent call first):
  /home/zhoubin/code/m88star5_work/m88star5-sdk/drivers/serial/atcuart100/CMakeLists.txt:3 (zephyr_library_sources)
```
and the previous 'No SOURCES given' alarm still exists.
```
CMake Warning at /home/zhoubin/code/m88star5_work/zephyr/CMakeLists.txt:864 (message):
  No SOURCES given to Zephyr library: drivers__serial

  Excluding target from build.
```
The directory structure of my driver is as follows：
```
drivers
├── CMakeLists.txt
├── Kconfig
└── serial
    ├── atcuart100
    │   ├── CMakeLists.txt
    │   ├── Kconfig
    │   └── mt_uart_atcuart100.c
    ├── CMakeLists.txt
    └── Kconfig
```
In drivers/serial/atcuart100/CMakeLists.txt, the code is as follows:
```
zephyr_library_amend()
zephyr_library_sources(mt_uart_atcuart100.c)
```
I need your help to solve this problem.

**Reply by gramsay0** [reply]
`zephyr_library_amend()` needs to be in drivers/serial/CMakeLists.txt rather than drivers/serial/atcuart100/CMakeLists.txt.

See the comment about `zephyr_library_amend()` in the link:
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/extensions.cmake#L461
```
# Note, in order to ensure correct library when amending, the folder structure in the
# Zephyr module must resemble the structure used in Zephyr, as example:
#
# Example: to amend the zephyr library created in
# ZEPHYR_BASE/drivers/entropy/CMakeLists.txt
# add the following file:
# ZEPHYR_MODULE/drivers/entropy/CMakeLists.txt
# with content:
# zephyr_library_amend()
# zephyr_library_sources(...)
```

edit: Also consider removing drivers/serial/atcuart100/CMakeLists.txt and adding the sources directly from drivers/serial/CMakeLists.txt. The closer you follow the existing pattern the less issues you will have with this

**Reply by foxriver1025** [reply]
Thank you very much, I have solved this problem after modifying it
